File: nestack-backend/src/common/utils/invite-code.util.ts
const CHARACTERS = 'ABCDEFGHJKLMNPQRSTUVWXYZ23456789'; // Removed similar looking characters

export function generateInviteCode(length = 8): string {
  let code = '';
  for (let i = 0; i < length; i++) {
    code += CHARACTERS.charAt(Math.floor(Math.random() * CHARACTERS.length));
  }
  return code;
}

export function formatInviteCode(code: string): string {
  // Format as XXXX-XXXX
  if (code.length === 8) {
    return `${code.slice(0, 4)}-${code.slice(4)}`;
  }
  return code;
}

export function normalizeInviteCode(code: string): string {
  // Remove dashes and convert to uppercase
  return code.replace(/-/g, '').toUpperCase();
}

File: nestack-backend/src/main/kotlin/com/nestack/common/util/InviteCodeUtil.kt
package com.nestack.common.util

import com.nestack.common.constant.Defaults

object InviteCodeUtil {
    // Exclude similar-looking characters: 0/O, 1/I/L
    private const val INVITE_CODE_CHARSET = "ABCDEFGHJKMNPQRSTUVWXYZ23456789"

    /**
     * Generate a 12-character invite code
     * Uses a character set that excludes confusing characters (0/O, 1/I/L)
     * @returns 12-character invite code (e.g., "A3B4C5D6E7F8")
     */
    fun generate(): String {
        return CryptoUtil.generateRandomString(Defaults.INVITE_CODE_LENGTH, INVITE_CODE_CHARSET)
    }

    /**
     * Validate invite code format
     * @param code Invite code to validate
     * @returns true if valid format
     */
    fun isValidFormat(code: String): Boolean {
        if (code.isBlank() || code.length != Defaults.INVITE_CODE_LENGTH) {
            return false
        }
        
        return code.all { it in INVITE_CODE_CHARSET }
    }

    /**
     * Normalize invite code (uppercase, trim)
     * @param code Invite code to normalize
     * @returns Normalized invite code
     */
    fun normalize(code: String): String {
        return code.trim().uppercase()
    }

    /**
     * Format invite code for display (add dashes for readability)
     * @param code 12-character invite code
     * @returns Formatted code (e.g., "A3B4-C5D6-E7F8")
     */
    fun formatForDisplay(code: String): String {
        if (code.length != Defaults.INVITE_CODE_LENGTH) {
            return code
        }
        return "${code.slice(0..3)}-${code.slice(4..7)}-${code.slice(8..11)}"
    }

    /**
     * Parse formatted invite code (remove dashes)
     * @param formattedCode Formatted invite code (may contain dashes)
     * @returns Raw 12-character invite code
     */
    fun parseFormattedCode(formattedCode: String): String {
        return formattedCode.replace("-", "").uppercase()
    }
}
